Class Cast exception from DefaultProjectDependencyMetadata to ModuleDependencyMetadata when using includeBuild and enforcedPlatform

### Summary:
When connecting the gradle project b to the project a with a dependency and also using a enforcedPlatform dependency to determine/enforce the version numbers of the dependent project, everything works fine, when both projects are NOT in a composite build. Once you try to add a `includeBuild '../project_a'` in the settings.gradle of project b, you will run into the class cast exception from DefaultProjectDependencyMetadata to ModuleDependencyMetadata. You can resolve this problem by adding another `includeBuild '../platform_project'` to the composite build, where you point to the corresponding gradle platform_project, but this is not wanted.

### Expected Behavior
It should be possible to use a enforcedPlatform dependency at the same time in the context of a composite build (without being forced to includeBuild also the gradle platform project)

### Current Behavior
When the two project are connected via an includeBuild entry in the settings.gradle file, you will get this exception when starting a gradlew build:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':Common:compileJava'.
> Could not resolve all dependencies for configuration ':Common:compileClasspath'.
   > org.gradle.internal.component.local.model.DefaultProjectDependencyMetadata cannot be cast to org.gradle.i
nternal.component.external.model.ModuleDependencyMetadata
```

### Context
We are trying to use the enforcedPlatform mechanism to use our company BOM file for the version-numbers of our dependency artifacts. At the same time we must be able to use the composite Build feature with eclipse buildShip plugin to create a good workspace for our developers. This will not work, when it is not possible to use enforcedPlatform at the same time with includeBuild. It also makes no sense to force all developers to check out at the same time the platform project when the want to use composite builds.

### Steps to Reproduce 
[forcedPlatformIncludeBuildProblem.zip](https://github.com/gradle/gradle/files/3635546/forcedPlatformIncludeBuildProblem.zip)
Extact all included folders and put them next to each other. Then go into the folder ci_poc_gitflow_platform and first publish with `gradlew pTML` the platform project to the local .m2. Afterwards use `pTML` task on the ci_poc_gitflow_a folder, to publish the a project. Now you will get the class cast exception in the ci_poc_gitflow_b folder when starting gradle with `gradlew build`. 
You can "fix" it, by commenting out the `includeBuild '../ci_poc_gitflow_a'` in the settings.gradle of root-project ci_poc_gitflow_b (=no composite build) or by activating the line `//includeBuild '../ci_poc_gitflow_platform'` to include the platform (=composite build with the gradle platform project)

### Your Environment
https://gradle.com/s/te7vykvm6lgco

I hope this information is enought. Please come back if you need more.

--- a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/platforms/EnforcedPlatformIntegrationTest.groovy
+++ b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/platforms/EnforcedPlatformIntegrationTest.groovy
@@ -28,6 +28,46 @@ class EnforcedPlatformIntegrationTest extends AbstractHttpDependencyResolutionTe
         """
     }
 
+    def "dependency substitution on a dependency enforced by a published platform should not cause an exception"() {
+        settingsFile << """
+            includeBuild 'jackson-core'
+        """
+        file('jackson-core/build.gradle') << """
+            plugins { id 'java-library' }
+            group = 'com.fasterxml.jackson.core'
+            version = '2.12.3-local-patch'
+        """
+
+        buildFile << """
+            apply plugin: 'java-library'
+
+            ${mavenCentralRepository()}
+
+            dependencies {
+                api enforcedPlatform('com.fasterxml.jackson:jackson-bom:2.12.3') // This BOM does not have module metadata (!)
+                api 'com.fasterxml.jackson.core:jackson-core'
+            }
+        """
+
+        def resolve = new ResolveTestFixture(buildFile, 'runtimeClasspath')
+        resolve.expectDefaultConfiguration('runtimeElements')
+        resolve.prepare()
+
+        when:
+        succeeds ':checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(':', ':test:') {
+                module('com.fasterxml.jackson:jackson-bom:2.12.3:enforced-platform-runtime') {
+                    noArtifacts()
+                    constraint('com.fasterxml.jackson.core:jackson-core:2.12.3', 'project :jackson-core', 'com.fasterxml.jackson.core:jackson-core:2.12.3-local-patch')
+                }
+                edge('com.fasterxml.jackson.core:jackson-core', 'project :jackson-core', 'com.fasterxml.jackson.core:jackson-core:2.12.3-local-patch') {}
+            }
+        }
+    }
+
     @ToBeFixedForConfigurationCache(because = "Resolve test fixture doesn't support configuration cache")
     def "dependency on unsatisfiable range shouldn't trigger null pointer exception"() {
         settingsFile << """

--- a/subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ForcedDependencyMetadataWrapper.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/ForcedDependencyMetadataWrapper.java
@@ -21,6 +21,7 @@ import org.gradle.api.artifacts.component.ModuleComponentSelector;
 import org.gradle.api.capabilities.Capability;
 import org.gradle.api.internal.attributes.AttributesSchemaInternal;
 import org.gradle.api.internal.attributes.ImmutableAttributes;
+import org.gradle.internal.component.local.model.DefaultProjectDependencyMetadata;
 import org.gradle.internal.component.model.ComponentResolveMetadata;
 import org.gradle.internal.component.model.ConfigurationMetadata;
 import org.gradle.internal.component.model.DependencyMetadata;
@@ -75,12 +76,20 @@ public class ForcedDependencyMetadataWrapper implements ForcingDependencyMetadat
 
     @Override
     public DependencyMetadata withTarget(ComponentSelector target) {
-        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) delegate.withTarget(target));
+        DependencyMetadata dependencyMetadata = delegate.withTarget(target);
+        if (dependencyMetadata instanceof DefaultProjectDependencyMetadata) {
+            return ((DefaultProjectDependencyMetadata) dependencyMetadata).forced();
+        }
+        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) dependencyMetadata);
     }
 
     @Override
     public DependencyMetadata withTargetAndArtifacts(ComponentSelector target, List<IvyArtifactName> artifacts) {
-        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) delegate.withTargetAndArtifacts(target, artifacts));
+        DependencyMetadata dependencyMetadata = delegate.withTargetAndArtifacts(target, artifacts);
+        if (dependencyMetadata instanceof DefaultProjectDependencyMetadata) {
+            return ((DefaultProjectDependencyMetadata) dependencyMetadata).forced();
+        }
+        return new ForcedDependencyMetadataWrapper((ModuleDependencyMetadata) dependencyMetadata);
     }
 
     @Override

--- a/subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultProjectDependencyMetadata.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/internal/component/local/model/DefaultProjectDependencyMetadata.java
@@ -25,13 +25,14 @@ import org.gradle.internal.component.model.ComponentResolveMetadata;
 import org.gradle.internal.component.model.ConfigurationMetadata;
 import org.gradle.internal.component.model.DependencyMetadata;
 import org.gradle.internal.component.model.ExcludeMetadata;
+import org.gradle.internal.component.model.ForcingDependencyMetadata;
 import org.gradle.internal.component.model.IvyArtifactName;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 
-public class DefaultProjectDependencyMetadata implements DependencyMetadata {
+public class DefaultProjectDependencyMetadata implements ForcingDependencyMetadata {
     private final ProjectComponentSelector selector;
     private final DependencyMetadata delegate;
     private final boolean isTransitive;
@@ -108,4 +109,19 @@ public class DefaultProjectDependencyMetadata implements DependencyMetadata {
         return delegate.withReason(reason);
     }
 
+    @Override
+    public boolean isForce() {
+        if (delegate instanceof ForcingDependencyMetadata) {
+            return ((ForcingDependencyMetadata) delegate).isForce();
+        }
+        return false;
+    }
+
+    @Override
+    public ForcingDependencyMetadata forced() {
+        if (delegate instanceof ForcingDependencyMetadata) {
+            return ((ForcingDependencyMetadata) delegate).forced();
+        }
+        return this;
+    }
 }

